In [38]:
import pandas as pd

In [39]:
test_2018 = pd.read_csv('../know_Data/test/KNOW_2018_test 밀린부분 복원.csv', encoding='cp949')

In [40]:
test_2018

idx  cq1  cq2  cq3  cq4  cq5  cq6  cq7  cq8  cq9  ...  bq37  \
0      9486    4    1    3    1    1    3    2    3    1  ...     1   
1      9487    5    4    5    5    5    5    3    3    3  ...     4   
2      9488    3    3    3    4    3    3    4    4    2  ...     4   
3      9489    4    4    4    4    2    3    3    4    4  ...     3   
4      9490    5    2    3    2    4    4    4    4    3  ...     2   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   
9064  18564    5    4    2    2    2    4    4    3    3  ...     4   
9065  18565    2    2    5    5    5    4    4    4    4  ...     5   
9066  18566    4    4    3    2    2    3    4    4    4  ...     4   
9067  18567    4    2    5    5    5    5    3    4    3  ...     5   
9068  18568    4    2    5    4    5    5    4    5    4  ...     4   

             bq37_1  bq38  bq38_1  bq38_2  bq39  bq40  bq41_1  bq41_2  bq41_3  
0                       2               6        35.0                    2000  
1     커뮤니케이션 디자인 전공     1       1             1  65.0    2700    2200          
2              국문학과     2               6        60.0                    3200  
3              극작학과     2               6        20.0                    2000  
4                실업     1       4             2  35.0    3000    1500          
...             ...   ...     ...     ...   ...   ...     ...     ...     ...  
9064          호텔조리학     2               5        72.0                    8000  
9065           식품공학     1       1             1  40.0    3000    2100          
9066           환경공학     1       2             2  40.0    1800    1800          
9067           식품공학     1       1             1  52.0    3500    2800          
9068             농학     1       1             1  40.0    2500    2500          

[9069 rows x 140 columns]

In [41]:
train_2018 = pd.read_csv('know_2018_enutf.csv')

In [42]:
test_2018.isnull().sum()[:50]

idx     0
cq1     0
cq2     0
cq3     0
cq4     0
cq5     0
cq6     0
cq7     0
cq8     0
cq9     0
cq10    0
cq11    0
cq12    0
cq13    0
cq14    0
cq15    0
cq16    0
cq17    0
cq18    0
cq19    0
cq20    0
cq21    0
cq22    0
cq23    0
cq24    0
cq25    0
cq26    0
cq27    0
cq28    0
cq29    0
cq30    0
cq31    0
cq32    0
cq33    0
cq34    0
cq35    0
cq36    0
cq37    0
cq38    0
cq39    0
cq40    0
cq41    0
cq42    0
cq43    0
cq44    0
cq45    0
cq46    0
cq47    0
cq48    0
cq49    0
dtype: int64

In [43]:
test_2018.isnull().sum()[50:100]

cq50_1    0
cq50_2    0
cq50_3    0
cq50_4    0
cq50_5    0
cq50_6    0
cq50_7    0
cq50_8    0
iq1       0
iq2       0
iq3       0
iq4       0
iq5       0
iq6       0
bq1       0
bq2       0
bq3       0
bq4       0
bq4_1a    0
bq4_1b    0
bq4_1c    0
bq5       0
bq5_1     0
bq5_2     0
bq6       0
bq7       0
bq8_1     0
bq8_2     0
bq8_3     0
bq9       0
bq10      0
bq11      0
bq12_1    0
bq12_2    0
bq12_3    0
bq12_4    0
bq12_5    0
bq13      0
bq14      0
bq15      0
bq16      0
bq17      0
bq18      0
bq19      0
bq20      0
bq21      0
bq221     0
bq222     0
bq223     0
bq231     0
dtype: int64

In [44]:
test_2018.isnull().sum()[100:]

bq232       0
bq233       0
bq234       0
bq235       0
bq241       0
bq242       0
bq243       0
bq244       0
bq245       0
bq25        0
bq25_1      0
bq26_1      0
bq26_1a     0
bq26_2      0
bq26_2a     0
bq26_3      0
bq26_3a     0
bq26_4      0
bq26_4a     0
bq27        0
bq28        0
bq28_1      0
bq29        0
bq30        0
bq31        1
bq32        0
bq33        0
bq34        0
bq35        0
bq36        0
bq37        0
bq37_1      0
bq38        0
bq38_1      0
bq38_2      0
bq39        0
bq40        0
bq41_1      0
bq41_2      0
bq41_3     32
dtype: int64

In [45]:
test_2018.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9069 entries, 0 to 9068
Data columns (total 140 columns):
 #    Column   Dtype  
---   ------   -----  
 0    idx      int64  
 1    cq1      int64  
 2    cq2      int64  
 3    cq3      int64  
 4    cq4      int64  
 5    cq5      int64  
 6    cq6      int64  
 7    cq7      int64  
 8    cq8      int64  
 9    cq9      int64  
 10   cq10     int64  
 11   cq11     int64  
 12   cq12     int64  
 13   cq13     int64  
 14   cq14     int64  
 15   cq15     int64  
 16   cq16     int64  
 17   cq17     int64  
 18   cq18     int64  
 19   cq19     int64  
 20   cq20     int64  
 21   cq21     int64  
 22   cq22     int64  
 23   cq23     int64  
 24   cq24     int64  
 25   cq25     int64  
 26   cq26     int64  
 27   cq27     int64  
 28   cq28     int64  
 29   cq29     int64  
 30   cq30     int64  
 31   cq31     int64  
 32   cq32     int64  
 33   cq33     int64  
 34   cq34     int64  
 35   cq35     int64  
 36   cq36     int

In [46]:
for i in test_2018.columns : 
    print(i, len(test_2018[i].value_counts()))

idx 9069
cq1 5
cq2 5
cq3 5
cq4 5
cq5 5
cq6 5
cq7 5
cq8 5
cq9 5
cq10 5
cq11 5
cq12 5
cq13 5
cq14 5
cq15 5
cq16 5
cq17 5
cq18 5
cq19 5
cq20 5
cq21 5
cq22 5
cq23 5
cq24 5
cq25 5
cq26 5
cq27 5
cq28 5
cq29 5
cq30 5
cq31 5
cq32 5
cq33 5
cq34 5
cq35 5
cq36 5
cq37 5
cq38 5
cq39 5
cq40 5
cq41 5
cq42 5
cq43 5
cq44 5
cq45 5
cq46 5
cq47 3
cq48 2
cq49 5
cq50_1 3
cq50_2 3
cq50_3 3
cq50_4 3
cq50_5 3
cq50_6 3
cq50_7 3
cq50_8 3
iq1 5
iq2 5
iq3 5
iq4 5
iq5 5
iq6 5
bq1 20
bq2 6
bq3 6
bq4 2
bq4_1a 1536
bq4_1b 738
bq4_1c 318
bq5 2
bq5_1 7
bq5_2 3860
bq6 7
bq7 7
bq8_1 5
bq8_2 5
bq8_3 5
bq9 5
bq10 5
bq11 5
bq12_1 5
bq12_2 6
bq12_3 6
bq12_4 6
bq12_5 5
bq13 3
bq14 2
bq15 4
bq16 2
bq17 3
bq18 3
bq19 3
bq20 5
bq21 49
bq221 4
bq222 4
bq223 2
bq231 6
bq232 4
bq233 3
bq234 2
bq235 2
bq241 6
bq242 2
bq243 2
bq244 2
bq245 2
bq25 3
bq25_1 3
bq26_1 3
bq26_1a 4
bq26_2 3
bq26_2a 4
bq26_3 3
bq26_3a 4
bq26_4 3
bq26_4a 4
bq27 2
bq28 5
bq28_1 7624
bq29 1684
bq30 5810
bq31 1822
bq32 2307
bq33 685
bq34 5
bq35 2
bq36 65
bq37 6


In [47]:
test_2018.bq39.value_counts()

1    6620
     1331
2    1118
Name: bq39, dtype: int64

1. 결측치

- 주관식 
bq31        1

- 처리해야하는 col
bq41_3     32

2. 이상치 

3. 주관식

- 자격증
bq4_1a 1536
bq4_1b 738
bq4_1c 318


- 버림
bq5_2 3860

bq28_1 7624
bq29 1684
bq30 5810
bq31 1822
bq32 2307
bq33 685

- 학과 
bq37_1 1867

4. 숫자여야하는 col
bq21 49
bq36 65
bq40 74
bq41_1 258
bq41_2 174
bq41_3 102

5. 체크해봐야하는 col

bq221 4
bq222 4
bq223 2
bq231 6
bq232 4
bq233 3
bq234 2
bq235 2
bq241 6
bq242 2
bq243 2
bq244 2
bq245 2

## 이상치 처리

In [48]:
for col in test_2018:
        test_2018[col].replace(' ', 0, inplace = True)

In [49]:
str_li =['bq4_1a','bq4_1b','bq4_1c','bq5_2','bq28_1','bq29','bq30','bq31', 'bq32','bq33','bq37_1']


In [50]:
test_2018 = test_2018.fillna(0)

In [51]:
test_2018.isnull().sum()[100:]

bq232      0
bq233      0
bq234      0
bq235      0
bq241      0
bq242      0
bq243      0
bq244      0
bq245      0
bq25       0
bq25_1     0
bq26_1     0
bq26_1a    0
bq26_2     0
bq26_2a    0
bq26_3     0
bq26_3a    0
bq26_4     0
bq26_4a    0
bq27       0
bq28       0
bq28_1     0
bq29       0
bq30       0
bq31       0
bq32       0
bq33       0
bq34       0
bq35       0
bq36       0
bq37       0
bq37_1     0
bq38       0
bq38_1     0
bq38_2     0
bq39       0
bq40       0
bq41_1     0
bq41_2     0
bq41_3     0
dtype: int64

In [52]:
for i in test_2018.columns : 
    if i not in str_li : 
        test_2018[i] = test_2018[i].astype(int)
    else :
        test_2018[i]=test_2018[i].astype(str)


In [53]:
test_2018.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9069 entries, 0 to 9068
Data columns (total 140 columns):
 #    Column   Dtype 
---   ------   ----- 
 0    idx      int32 
 1    cq1      int32 
 2    cq2      int32 
 3    cq3      int32 
 4    cq4      int32 
 5    cq5      int32 
 6    cq6      int32 
 7    cq7      int32 
 8    cq8      int32 
 9    cq9      int32 
 10   cq10     int32 
 11   cq11     int32 
 12   cq12     int32 
 13   cq13     int32 
 14   cq14     int32 
 15   cq15     int32 
 16   cq16     int32 
 17   cq17     int32 
 18   cq18     int32 
 19   cq19     int32 
 20   cq20     int32 
 21   cq21     int32 
 22   cq22     int32 
 23   cq23     int32 
 24   cq24     int32 
 25   cq25     int32 
 26   cq26     int32 
 27   cq27     int32 
 28   cq28     int32 
 29   cq29     int32 
 30   cq30     int32 
 31   cq31     int32 
 32   cq32     int32 
 33   cq33     int32 
 34   cq34     int32 
 35   cq35     int32 
 36   cq36     int32 
 37   cq37     int32 
 38   cq38  

In [54]:
# bq37=1(중졸)인 사람중 학과 입력한 사람 -> '0'으로 처리 
test_2018[(test_2018.bq37==1) & (test_2018.bq37_1!='0')]['bq37_1'].value_counts()

컴퓨터공학    1
Name: bq37_1, dtype: int64

In [55]:
test_2018[(test_2018.bq37==1) & (test_2018.bq37_1!='0')].index

Int64Index([2686], dtype='int64')

In [56]:
# 학과 입력했기 때문에 최종학력이 고졸이상으로 추정됨
# 수기입력의 오류라고 생각되어 문제지에서 1과 가까운 3으로 대체 

test_2018.iloc[2686,130]=3

In [57]:
test_2018[(test_2018.bq37==1) & (test_2018.bq37_1!='0')].index

Int64Index([], dtype='int64')

In [58]:
# 고졸 이상중 학과 입력안한 사람 
# 앞에서 주관식 항목의 빈값(결측치)를 '0'으로 처리했기 때문에 '0'으로 찾아야함
# 결측치 '0'으로 처리
test_2018[(test_2018.bq37.isin([2,3,4,5,6])) & (test_2018.bq37_1=='0')]['bq37'].value_counts()

2    110
4     24
3     11
5      5
Name: bq37, dtype: int64

In [59]:
# 근로시간
# 결측치 없음
test_2018[test_2018.bq40==0].index

Int64Index([], dtype='int64')

In [60]:
## 근로소득 처리 
# test 결측치를 train의 업종의 중위값으로 대체 

In [61]:
# 임금근로자

null_li = test_2018[(test_2018.bq38==1) & (test_2018.bq41_1==0)].index
null_li

Int64Index([962, 1524, 1799, 2708, 3070, 7609, 8070], dtype='int64')

In [62]:
for i in null_li:     
    test_2018.iloc[i,137]=round(train_2018[(train_2018.bq1 ==test_2018.iloc[i,64])&(train_2018.bq41_1!=0)].bq41_1.median())

In [63]:
null_li = test_2018[(test_2018.bq38==1) & (test_2018.bq41_1==0)].index
null_li

Int64Index([], dtype='int64')

In [64]:
# 임금근로자 초봉
null_li2 = test_2018[(test_2018.bq38==1) & (test_2018.bq41_2==0)].index
null_li2

Int64Index([  59,  141,  152,  331,  720,  783,  808,  860,  962,  966, 1157,
            1367, 1397, 1524, 1578, 1648, 1744, 1799, 1835, 1922, 2027, 2070,
            2073, 2234, 2440, 2478, 2708, 2917, 3070, 3099, 3169, 3291, 3302,
            3524, 3758, 3772, 3944, 4002, 4101, 4112, 4122, 4143, 4171, 4305,
            4418, 4498, 4686, 4689, 4786, 4864, 4897, 4971, 4999, 5070, 5078,
            5252, 5270, 5318, 5505, 5572, 5849, 5931, 6177, 6451, 6460, 6476,
            6605, 6607, 7011, 7128, 7236, 7299, 7351, 7362, 7609, 7656, 7670,
            7740, 7967, 7984, 8064, 8070, 8231, 8323, 8478, 8479, 8482, 8607,
            8714, 8722, 8760, 8780, 8820, 8929, 8932, 8995, 9030],
           dtype='int64')

In [65]:
for i in null_li2:     
    test_2018.iloc[i,138]=round(train_2018[(train_2018.bq1 ==test_2018.iloc[i,64])&(train_2018.bq41_2!=0)].bq41_2.median())

In [66]:
null_li2 = test_2018[(test_2018.bq38==1) & (test_2018.bq41_2==0)].index
null_li2

Int64Index([], dtype='int64')

In [67]:
# 비임금근로자 근로소득

null_li3 = test_2018[(test_2018.bq38==2) & (test_2018.bq41_3==0)].index
null_li3

Int64Index([  34,  661, 1751, 1875, 2193, 2488, 2688, 4403, 4516, 4558, 4768,
            4914, 4936, 4943, 4947, 5487, 6840, 6925, 7995, 8349, 8428, 8623,
            8922],
           dtype='int64')

In [68]:
for i in null_li3:     
    test_2018.iloc[i,139]=round(train_2018[(train_2018.bq1 ==test_2018.iloc[i,64])&(train_2018.bq41_3!=0)].bq41_3.median())

In [69]:
null_li3 = test_2018[(test_2018.bq38==2) & (test_2018.bq41_3==0)].index
null_li3

Int64Index([], dtype='int64')

In [ ]:
# 18년 자격증 주관식에는 .값을 '0'으로 처리 

In [70]:
## 자격증 없음 

for i in test_2018.bq4_1b:
    if '.' in i : 
        print(i)

test_2018['bq4_1b'].replace('.', '0', inplace = True)

.


In [71]:
for i in test_2018.bq4_1b:
    if '.' in i : 
        print(i)

In [72]:


for i in test_2018.bq4_1c:
    if '.' in i : 
        print(i)

test_2018['bq4_1c'].replace('.', '0', inplace = True)

.
간접투자.증권투자자격증


In [73]:
for i in test_2018.bq4_1c:
    if '.' in i : 
        print(i)

간접투자.증권투자자격증


In [74]:
test_2018.to_csv("know_2018_test_enutf.csv", encoding='utf8', index=False)
test_2018.to_csv("know_2018_test_encp.csv", encoding='cp949', index=False)